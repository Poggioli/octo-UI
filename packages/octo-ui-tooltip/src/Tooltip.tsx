import { forwardRef } from "@octo-ui/core";
import {
  StyledTooltipArrow,
  StyledTooltipContent,
  StyledTooltipPortal,
  StyledTooltipProvider,
  StyledTooltipRoot,
  StyledTooltipTrigger,
} from "./styles";
import {
  TooltipArrowProps,
  TooltipContentProps,
  TooltipPortalProps,
  TooltipProps,
  TooltipProviderProps,
  TooltipTriggerProps,
} from "./types";

const TooltipProvider = forwardRef<
  typeof StyledTooltipProvider,
  TooltipProviderProps
>(({ ...props }, forwardedRef) => {
  return <StyledTooltipProvider {...props} ref={forwardedRef} />;
});

// ========================================================================= //

const Tooltip = forwardRef<typeof StyledTooltipRoot, TooltipProps>(
  (
    {
      children,
      delayDuration,
      skipDelayDuration,
      disableHoverableContent,
      ...props
    },
    forwardedRef
  ) => {
    return (
      <TooltipProvider
        delayDuration={delayDuration}
        skipDelayDuration={skipDelayDuration}
        disableHoverableContent={disableHoverableContent}
      >
        <StyledTooltipRoot {...props} ref={forwardedRef}>
          {children}
        </StyledTooltipRoot>
      </TooltipProvider>
    );
  }
);
Tooltip.toString = () => `.${StyledTooltipRoot.className}`;

// ========================================================================= //

const TooltipTrigger = forwardRef<
  typeof StyledTooltipTrigger,
  TooltipTriggerProps
>(({ ...props }, forwardedRef) => {
  return <StyledTooltipTrigger {...props} ref={forwardedRef} asChild />;
});
TooltipTrigger.toString = () => `.${StyledTooltipTrigger.className}`;

// ========================================================================= //

const TooltipPortal = forwardRef<
  typeof StyledTooltipPortal,
  TooltipPortalProps
>(({ ...props }, forwardedRef) => {
  return <StyledTooltipPortal {...props} ref={forwardedRef} />;
});

// ========================================================================= //

const TooltipArrow = forwardRef<typeof StyledTooltipArrow, TooltipArrowProps>(
  ({ ...props }, forwardedRef) => {
    return <StyledTooltipArrow {...props} ref={forwardedRef} />;
  }
);

// ========================================================================= //

const TooltipContent = forwardRef<
  typeof StyledTooltipContent,
  TooltipContentProps
>(
  (
    {
      children,
      forceMountPortal,
      container,
      height,
      width,
      collisionPadding = 20,
      sideOffset = 5,
      ...props
    },
    forwardedRef
  ) => {
    return (
      <TooltipPortal forceMount={forceMountPortal} container={container}>
        <StyledTooltipContent
          {...props}
          ref={forwardedRef}
          collisionPadding={collisionPadding}
          sideOffset={sideOffset}
        >
          {children}
          <TooltipArrow height={height} width={width} />
        </StyledTooltipContent>
      </TooltipPortal>
    );
  }
);
TooltipContent.toString = () => `.${StyledTooltipContent.className}`;

export {
  StyledTooltipArrow,
  StyledTooltipContent,
  StyledTooltipPortal,
  StyledTooltipProvider,
  StyledTooltipRoot,
  StyledTooltipTrigger,
};
export { Tooltip, TooltipTrigger, TooltipContent };
