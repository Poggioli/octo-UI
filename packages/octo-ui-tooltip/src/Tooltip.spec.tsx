import { fireEvent, render } from "@testing-library/react";
import { Tooltip, TooltipContent, TooltipTrigger } from "./Tooltip";

describe("<Tooltip />", () => {
  it(`GIVEN a <Tooltip />
      WHEN render is called
      THEN should render a <Tooltip />`, () => {
    const rendered = render(
      <Tooltip>
        <TooltipTrigger>
          <button>Trigger</button>
        </TooltipTrigger>
        <TooltipContent>
          <p>Tooltip content</p>
        </TooltipContent>
      </Tooltip>
    );
    expect(rendered.container).toMatchSnapshot();
  });

  it(`GIVEN a <Tooltip />
      WHEN trigger is hovered
      THEN should render a tooltip role`, () => {
    const rendered = render(
      <Tooltip>
        <TooltipTrigger>
          <button>Trigger</button>
        </TooltipTrigger>
        <TooltipContent>
          <p>Tooltip content</p>
        </TooltipContent>
      </Tooltip>
    );

    const triggerElement = rendered.getByText("Trigger");
    fireEvent.focus(triggerElement);

    const contentElement = rendered.getByRole("tooltip");
    expect(contentElement).toBeVisible();
    expect(rendered.baseElement).toMatchSnapshot();
  });

  it(`GIVEN a <Tooltip />
      WHEN toString is called
      THEN return a string`, () => {
    expect(Tooltip.toString()).toBe(".octo-UI-c-PJLV");
  });

  it(`GIVEN a <TooltipTrigger />
      WHEN toString is called
      THEN return a string`, () => {
    expect(TooltipTrigger.toString()).toBe(".octo-UI-c-PJLV");
  });

  it(`GIVEN a <TooltipContent />
      WHEN toString is called
      THEN return a string`, () => {
    expect(TooltipContent.toString()).toBe(".octo-UI-c-cdrPBc");
  });
});
