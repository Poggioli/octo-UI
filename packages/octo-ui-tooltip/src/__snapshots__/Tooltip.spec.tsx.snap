// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> GIVEN a <Tooltip />
      WHEN render is called
      THEN should render a <Tooltip /> 1`] = `
<div>
  <button
    class="octo-UI-c-PJLV"
    data-state="closed"
  >
    Trigger
  </button>
</div>
`;

exports[`<Tooltip /> GIVEN a <Tooltip />
      WHEN trigger is hovered
      THEN should render a tooltip role 1`] = `
<body>
  <div>
    <button
      aria-describedby="radix-:r1:"
      class="octo-UI-c-PJLV"
      data-state="instant-open"
    >
      Trigger
    </button>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    style="position: fixed; left: 0px; top: 0px; transform: translate3d(0, -200%, 0); min-width: max-content;"
  >
    <div
      class="octo-UI-c-cdrPBc"
      data-align="center"
      data-side="top"
      data-state="instant-open"
      style="--radix-tooltip-content-transform-origin: var(--radix-popper-transform-origin); --radix-tooltip-content-available-width: var(--radix-popper-available-width); --radix-tooltip-content-available-height: var(--radix-popper-available-height); --radix-tooltip-trigger-width: var(--radix-popper-anchor-width); --radix-tooltip-trigger-height: var(--radix-popper-anchor-height); animation: none;"
    >
      <p>
        Tooltip content
      </p>
      <span
        style="position: absolute; bottom: 0px; transform: translateY(100%); visibility: hidden;"
      >
        <svg
          class="octo-UI-c-cKfwvy"
          height="5"
          preserveAspectRatio="none"
          style="display: block;"
          viewBox="0 0 30 10"
          width="10"
        >
          <polygon
            points="0,0 30,0 15,10"
          />
        </svg>
      </span>
      <span
        id="radix-:r1:"
        role="tooltip"
        style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      >
        <p>
          Tooltip content
        </p>
      </span>
    </div>
  </div>
</body>
`;
